### PR TITLE
feat(fxa-15): add "apps" to settings

### PR DESCRIPTION
**Pull request by vladikoff**
Connects to https://github.com/mozilla/fxa-features/issues/20

### CODE REVIEW

**Comment by rfk** [comment]
@vladikoff random thought - do you think it would make sense to move these feature cards into the fxa-features repo, and then PRs like this one can be attached to the feature issue in the product-planning waffleboard, rather than show up in the engineering waffleboard?

File: features/FxA-15-OAuth-Management/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+![](https://www.lucidchart.com/publicSegments/view/bcc7ac8f-cb6d-44cd-a5ca-b59a786168ca/image.jpeg?w=1)
+
+[Edit Mockup](https://www.lucidchart.com/documents/edit/bdafa87d-ea17-4beb-8381-f6da19d7359a#)
+
+The initial session management dashboard would
+revoke the non-expired tokens using the web UI.
+It will be displayed to logged in users of FxA
+in the settings view.
+
+It will also list the sessions that were created
```
Comment (shane-tomlinson) [inline features/FxA-15-OAuth-Management/README.md]: Does `sessions` here mean `OAuth tokens`? If so, we are going to have to be a whole lot more strict about deleting OAuth tokens on the client side, or giving them a short expiration as @rfk mentioned for another bug. We create OAuth codes/tokens in several places and leave many of them hanging around. I'm unsure how the browser deals with this.

One question that comes out of this, say I revoke a token for the browser. The browser could just then request a new one. Would revoking access to the browser, or FxA for that matter, prevent new tokens from being created?

**Comment by vladikoff** [comment]
> @vladikoff random thought - do you think it would make sense to move these feature cards into the fxa-features repo, and then PRs like this one can be attached to the feature issue in the product-planning waffleboard, rather than show up in the engineering waffleboard?

I think I'd prefer the feature cards to be closer to engineering, but we can discuss that in the 1:1.

**Comment by vladikoff** [comment]
@ryanfeeley could you please check the "Detailed design" section (and maybe other sections) and let me know what you think about strings and etc?

File: features/FxA-15-OAuth-Management/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+* The button to remove all active sessions will say "revoke access".
+* The dashboard will provide a "last authenticated" date in the same format as the devices view.
+* It will have an intro text that will say: "These are the services that can access your Firefox Account.".
+* The dashboard's title in settings will be "Services".
+* All services will have the same default service icon to improve the visual UX of the feature.
+* Revoking Access:
+  * OAuth service: will revoke all tokens belonging to a given OAuth service.
+  * FxA Settings (content-server): will invalidate all content-server tokens and local session token, log the user out.
+    * The content-server will be called "Firefox Accounts Settings" in the dashboard.
+  * Firefox Desktop and other browsers clients: will revoke the Firefox desktop OAuth tokens.
```
Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: From mtg: we should hide this somehow. mark browsers as 'a device client'

Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: from mtg: add a note - hide the thing behind the pref

File: features/FxA-15-OAuth-Management/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+services that I logged into.  
+
+****
+
+## Outcomes
+
+An OAuth management dashboard in the settings view.
+The success criteria of this feature:
+* 2%+ of users that view FxA Settings use the OAuth management dashboard.
+* at least 1% of users that viewed the OAuth management dashboard revoked a service.
```
Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: add `If you login to X then it does Y`. list of things that the feature does.

**Comment by vladikoff** [comment]
cc @ryanfeeley

**Comment by ryanfeeley** [comment]
@vladikoff  I don't think users will benefit from a section for oauth services and another for sync clients. I am hoping we can integrate them into one section, am just struggling with a good name.

From other websites, I see this:

FACEBOOK
* Your Browsers & Apps
* Where You’re Logged In
* Logged in with Facebook

GITHUB
* OAuth applications
* Security: Sessions

TWITTER
* Your Twitter data: Device history
* Your Twitter data: Login history
* Apps

APPLE ID
* Devices

Am going to compile a list of potential names.

**Comment by ryanfeeley** [comment]
@vladikoff Would this work?
connected-applications

**Comment by vladikoff** [comment]
> I am hoping we can integrate them into one section,

Sadly this is not possible. The tokens can be reused and there is no way to tell on which device.

**Comment by ryanfeeley** [comment]
@vladikoff I'm not sure I understand. You will have a list above Sync devices, and a list of OAuth applications. You put them together and sort by last active time. No?

**Comment by vladikoff** [comment]
@ryanfeeley oh sorry I got a bit confused (mostly because I saw the same device icon). That might work. any thoughts on this @rfk ? (also @ckarlof had some opinions on this :) )

**Comment by vladikoff** [comment]
@ryanfeeley I am a bit worried if "I'm a user and I want to disconnect my work Firefox profile", would I go and look for it in "Connected Applications". *shrugs*

**Comment by ryanfeeley** [comment]
@vladikoff If they think of "sync" as an application maybe? if users don't understand it (I can test this) we could also say "Connected apps & devices".

**Comment by vladikoff** [comment]
> @vladikoff If they think of "sync" as an application maybe? if users don't understand it (I can test this) we could also say "Connected apps & devices".

That works for me 👍 . I think I was only hoping you could have mentioned that we would squash the list together during the Vidyo meeting so we could all brainstorm that together.

**Comment by rfk** [comment]
> If they think of "sync" as an application maybe? 

I like where this train of thought is heading, will be interested to see a mockup but I think it could work

**Comment by ryanfeeley** [comment]
I'm trying some "generic" app icons, but my team says that because the list of supported services is small, we should ship per-service icons. How you engineer that is up to you, but the user should see our feature/product icon beside the feature/product name.

**Comment by rfk** [comment]
FWIW I also like the "apps and devices" phrasing as it will make it more clear to the user where to look for their device.

**Comment by vladikoff** [comment]
@ryanfeeley @rfk r? 

@ryanfeeley could you please provide the final mock up for the "Connected apps and devices view" ? I Think the current one has the wrong title in it.

**Comment by ryanfeeley** [comment]
@vladikoff The assumption @davismtl and I are making is that **Devices** will ship before **Devices & apps**. I'm working on getting **Devices** perfect first. Make sense?

**Comment by davismtl** [comment]
@vladikoff Our view is that we should iteratively ship this devices and apps rather than wait for it to be perfect and cover both. We were thinking devices first, apps/services second. It might also make measuring success easier.

Let us know how you feel about this.

**Comment by vladikoff** [comment]
> I'm working on getting Devices perfect first

@ryanfeeley that makes sense. I think I'm missing your point, what are you trying to say? Do you not want to give approval to proposed strings? or is not fully clear how `connected apps` will fit in?
We can maybe merge this feature proposal and follow up with a UX review before turning off the feature flag. *shrug*

> @vladikoff Our view is that we should iteratively ship this devices and apps rather than wait for it to be perfect and cover both. We were thinking devices first, apps/services second. It might also make measuring success easier.

The feature flag `&forceServiceList=1` can edit the title of the section from `Devices` -> `Connected apps and devices` and add apps to the list. If the flag is off then the devices view would not be changed.

**Comment by ryanfeeley** [comment]
@vladikoff This is the latest mockup which will soon be inside a feature doc. Until we have specific icons for OAuth services like AMO, the desktop icon will suffice. cc/ @davismtl 
devices-and-apps

**Comment by vladikoff** [comment]
> @vladikoff This is the latest mockup

Thanks! 👍

**Comment by rfk** [comment]
> The assumption @davismtl and I are making is that Devices
> will ship before Devices & apps. I'm working on getting Devices perfect first. Make sense?

Just to comment that I think this is a very sensible staging, as long as we have enough of a high-level idea of what the "Devices & apps" part will look like, that @vladikoff can push ahead with initial prototyping in the meantime.  It sounds like we are at that point, so 👍

File: features/FxA-15-OAuth-Management/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
+Disconnect OAuth services from FxA
+======================
+
+https://mozilla.aha.io/features/FXA-15
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: It's probably going to be more useful to link to e.g. https://github.com/mozilla/fxa-features/issues/20 rather than the soon-to-be-defunct Aha card.

File: features/FxA-15-OAuth-Management/README.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+* Being able to view logged in OAuth services in the apps view.
+* Being able to revoke access ("disconnect") a service.
+
+See "Detailed Design" below for more details.
+
+## Hypothesis
+
+We believe that by adding OAuth services and functionality to disconnect the services
+for Firefox Accounts users will improve the quality and security of FxA
+for everybody.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: There's some funky sentence structure going on here that makes this hard to parse, it might need a comma or two at least.

File: features/FxA-15-OAuth-Management/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+## Metrics
+
+We would collect usage metrics with StatsD:
+* number of views of the apps and devices view,
+* number of revoke actions for any service.
+* number of revoke actions for each service by client id.
+
+Derive a percentage of settings views vs apps view to understand
+the popularity of this feature.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: SGTM, and I think we should have two graphs to capture the metrics you outline above:

* Number of views the devices+apps screen, as a percentage of views of the settings page.
* Number of disconnects, as a percentage of views of the devices+apps screen.

I also wonder, does it make sense for us to link to this view from anywhere?  If so, adding inbound link tracking would be a valuable metric, but I'm not sure we plan to link to it anywhere that's oauth-specific.

File: features/FxA-15-OAuth-Management/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+****
+
+## Detailed design
+
+<img src='pr_moz_mock.png' width='450' />
+
+The initial OAuth apps list would
+revoke the non-expired tokens using the web UI.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: I suggest expanding this slightly, to "would show all non-expired tokens using the web UI, and allow the user ti revoke them".

File: features/FxA-15-OAuth-Management/README.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+It will be displayed to logged in users of FxA
+in the "Connected apps and devices" settings view.
+This new functionality changes the current devices view and adds to it.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: Do we care that the URL for this view will be https://accounts.firefox.com/settings/devices even though it has more than just devices in the list?  I think it's *probably* OK like that, but worth a thinking about.

File: features/FxA-15-OAuth-Management/README.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+* The dashboard will order the services by last login to a given service.
+* The button to remove all active sessions will say "Disconnect". To match "devices".
+* The dashboard will provide a "last active" date in the same format as the devices view.
+* The dashboard's title in settings will be "Connected apps and devices".
+* All services will have the same default service icon to improve the visual UX of the feature.
+* Revoking Access:
+  * OAuth service: will revoke all tokens belonging to a given OAuth service.
+  * FxA Settings (content-server): will invalidate all content-server tokens and local session token, log the user out.
+    * The content-server will be called "Firefox Accounts Settings" in the dashboard.
+  * Firefox Desktop and other browsers clients: will not be listed and their devices records would revoke their tokens.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: "will not be listed" -> "will stay the same as in the devices view feature"?

File: features/FxA-15-OAuth-Management/README.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+The initial OAuth apps list would
+revoke the non-expired tokens using the web UI.
+It will be displayed to logged in users of FxA
+in the "Connected apps and devices" settings view.
+This new functionality changes the current devices view and adds to it.
+
+The list will also include the sessions that were created
+for the Firefox Accounts website and the Firefox Desktop browser.
+This gives users full control over all of their sessions.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: I feel like there may be some interesting technical details here that we should call out.  Will we need to ask the auth-server for a list of all sessionTokens, display those that have device records as devices, and display the leftover ones as "Firefox Accounts website"?  We can't necessarily rely on the oauth server to give us a complete list here.

File: features/FxA-15-OAuth-Management/README.md (hunk 4 above)
Comment (ryanfeeley) [inline features/FxA-15-OAuth-Management/README.md]: > as long as we have enough of a high-level idea of what the "Devices & apps" part will look like

Mockup above is very close to final: https://github.com/mozilla/fxa/pull/171#issuecomment-238696839

**Comment by vladikoff** [comment]
Updated

@rfk @ryanfeeley @davismtl r?

Rendered: https://github.com/mozilla/fxa/blob/1589be27c01b0ed63aad24eab51f185b933b76fa/features/FxA-15-OAuth-Management/README.md
Feedback / Comment view: https://github.com/mozilla/fxa/pull/181/files

File: features/FxA-15-OAuth-Management/README.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+[Devices View (FXA-89)](https://github.com/mozilla/fxa/pull/181/files).
+
+The initial OAuth apps list would
+revoke the non-expired tokens using the web UI.
+It will be displayed to logged in users of FxA
+in the "Devices & apps" settings view.
+This new functionality changes the current devices view and adds to it.
+
+The list will also include the sessions that were created
+for the Firefox Accounts website. Firefox Browser and Sync logins
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: 👍 I'm really pro the idea of showing everything that's connected to the account in this view - devices, web logins, and oauth sessions.

File: features/FxA-15-OAuth-Management/README.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+* The dashboard's title in settings will be "Devices & apps".
+* All services will have the same default service icon to improve the visual UX of the feature.
+* Revoking Access:
+  * OAuth service: will revoke all non-expired tokens belonging to a given OAuth service.
+  * Only the OAuth services that do not need permission to get implicit grants will be listed (`canGrant=0`).
+  * Expired tokens are deleted by a different process. This feature does not deal with expired tokens.
+  * FxA Settings (content-server): will not be visible in this view.
+  * Firefox Desktop and other browsers clients: will not be listed and their devices records would revoke their tokens.
+* Initial version will show the default "Desktop" icon for the connected app.
+* The list will be hidden behind the flag in the initial version.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: It's not obvious what "the flag" refers to here. (well, I know what you mean, but it's not defined in the document)

File: features/FxA-15-OAuth-Management/README.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+* The button to remove all active sessions will say "Disconnect". To match "devices".
+* The dashboard will provide a "last active" date in the same format as the devices view.
+* The dashboard's title in settings will be "Devices & apps".
+* All services will have the same default service icon to improve the visual UX of the feature.
+* Revoking Access:
+  * OAuth service: will revoke all non-expired tokens belonging to a given OAuth service.
+  * Only the OAuth services that do not need permission to get implicit grants will be listed (`canGrant=0`).
+  * Expired tokens are deleted by a different process. This feature does not deal with expired tokens.
+  * FxA Settings (content-server): will not be visible in this view.
+  * Firefox Desktop and other browsers clients: will not be listed and their devices records would revoke their tokens.
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: What does "their devices records would revoke their tokens" mean?  Do we expect any OAuth tokens held by the device to be destroyed as part of disconnecting a device?

File: features/FxA-15-OAuth-Management/README.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+* The dashboard will order the services by last login to a given service.
+* The OAuth servers calls the listed items "clients". To make it easier to understand
+we call them "apps" in the settings view.
+* The button to remove all active sessions will say "Disconnect". To match "devices".
+* The dashboard will provide a "last active" date in the same format as the devices view.
+* The dashboard's title in settings will be "Devices & apps".
+* All services will have the same default service icon to improve the visual UX of the feature.
+* Revoking Access:
+  * OAuth service: will revoke all non-expired tokens belonging to a given OAuth service.
+  * Only the OAuth services that do not need permission to get implicit grants will be listed (`canGrant=0`).
```
Comment (rfk) [inline features/FxA-15-OAuth-Management/README.md]: I remember discussing this, but I don't recall what the reasoning was being excluding `canGrant=1` clients.  Is it because we expect them to show up as active sessionTokens in this view?

**Comment by rfk** [comment]
@vladikoff if you think it's valuable, I'd encourage to to add a list of acceptance criteria and maybe even a couple of "phases" in the style of https://github.com/mozilla/fxa/pull/181/files#diff-9d765e5787806afc80bdaa86191886a8R291

It's my hope that such a list will be useful for doing structured testing of the feature once it's shipped, to ensure we haven't missed any pieces.

File: features/FxA-15-OAuth-Management/README.md (hunk 14 above)
Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: >  Is it because we expect them to show up as active sessionTokens in this view?

Yeap exactly. Added this after chatting with @seanmonstar about it.

File: features/FxA-15-OAuth-Management/README.md (hunk 13 above)
Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: > Do we expect any OAuth tokens held by the device to be destroyed as part of disconnecting a device?

As part of some other feature maybe. It would be nice if it did that.

File: features/FxA-15-OAuth-Management/README.md (hunk 11 above)
Comment (vladikoff) [inline features/FxA-15-OAuth-Management/README.md]: > web logins, 

I'm removing this section. Adding a note to future improvements. With the 5 min TTL limit the content-server token disappears very quickly. 

Re: Firefox OAuth sessions -> https://github.com/mozilla/fxa/pull/171#discussion_r77062619

**Comment by rfk** [comment]
@vladikoff as discussed, this accurately documents what we're going to do for phase 1 of the feature, so Iet's merge it and iterate from there.
